### PR TITLE
pageserver: only store SLRUs & aux files on shard zero

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Since https://github.com/neondatabase/neon/pull/9423 the non-zero shards no longer need SLRU content in order to do GC.  This data is now redundant on shards >0.

One release cycle after merging that PR, we may merge this one, which also stops writing those pages to shards > 0, reaping the efficiency benefit.

Closes: https://github.com/neondatabase/neon/issues/7512
Closes: https://github.com/neondatabase/neon/issues/9641

## Summary of changes

- Avoid storing SLRUs on non-zero shards
- Bonus: avoid storing aux files on non-zero shards

### CODE REVIEW

File: pageserver/src/walingest.rs
Comment (VladLazar) [inline pageserver/src/walingest.rs:281]: I'm not sure about this.
`WalIngest::new` reads the checkpoint from storage and then uses it throughout. Looking at uses of `WalIngest::checkpoint`, I think we rely on having an up to date checkpoint.

---

Also, I don't see the need for this change. We could cut down on some records going to all pageservers if this is correct, but from my experiments said records don't add up to much and we are going to compress them anyway.

**Review by VladLazar — COMMENTED** [review]
Generally looks good. Unsure about the checkpoint change though.

**Comment by jcsp** [comment]
Rebased on main to de-conflict with decoder.rs changes.

Assuming it's all good, let's merge this on Monday after the release is cut, so that we get a full week to bake in staging.

File: libs/wal_decoder/src/decoder.rs
Comment (erikgrinaker) [inline libs/wal_decoder/src/decoder.rs:103]: In #9895, I added shard filtering below this block, instead of propagating it into the decode methods. This was mostly motivated by the fact that multiple decode methods may return the same metadata record type which needs common shard filtering logic, but it might be worthwhile to decouple them regardless.

Consider moving the shard filtering for logical messages below as well, for consistency. Alternatively, let's collapse the `heapam` and `neonmgr` logic and move shard filtering into those decode methods (I can do that as a followup).

File: pageserver/src/pgdatadir_mapping.rs
Comment (erikgrinaker) [inline pageserver/src/pgdatadir_mapping.rs:1473]: Why aren't these (debug) assertions?

**Review by erikgrinaker — APPROVED** [review]
Not very familiar with SLRUs so get another approval on this, but LGTM.

File: libs/pageserver_api/src/shard.rs
Comment (jcsp) [reply libs/pageserver_api/src/shard.rs:174]: (I don't have a lot of inspiration for making the naming right right now, but it can definitely be improved)

File: libs/pageserver_api/src/shard.rs
Comment (jcsp) [reply libs/pageserver_api/src/shard.rs:197]: We rely on compaction happening sooner or later to clean things up (apart from in shard splits, where there's a special step in compaction to filter layers) -- it's generally expensive to do it retroactively because one might have to download layers from S3.

File: pageserver/src/pgdatadir_mapping.rs
Comment (jcsp) [reply pageserver/src/pgdatadir_mapping.rs:1473]: Yeah, it should be an assertion.  Changed in bf07db61c5783c518dbb673773dafb35f460042f

**Pull request merged**
